src/main/java/org/example/homework03/service/UserServiceImpl.java

```diff
@@ -0,0 +1,48 @@
+package org.example.homework03.service;
+
+import org.example.homework03.domain.repository.UserRepository;
+import org.example.homework03.domain.service.UserService;
+import org.example.homework03.dto.UserRegistrationDto;
+import org.example.homework03.dto.UserResponseDto;
+import org.example.homework03.entity.User;
+import org.example.homework03.mapper.UserMapper;
+
+import java.util.Optional;
+
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+
+    public UserServiceImpl(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    public UserResponseDto registerUser(UserRegistrationDto userRegistrationDto) {
+        User user = userRepository.persistUser(UserMapper.toEntity(userRegistrationDto));
+
+        return UserMapper.toResponseDto(user);
+    }
+
+    public Optional<UserResponseDto> getUserById(Long userId) {
+        Optional<User> optionalUser = userRepository.getUserById(userId);
+
+         if (optionalUser.isEmpty()) {
+            return Optional.empty();
+         }
+
+         UserResponseDto userResponseDto = UserMapper.toResponseDto(optionalUser.get());
+
+         return Optional.of(userResponseDto);
+    }
```

Comment:
Refactor to throw EntityNotFoundException if optional is empty:
```suggestion
    public UserResponseDto getUserById(Long userId) {
        return userRepository.getUserById(userId)
            .map(UserMapper::toResponseDto) // Map the User entity to UserResponseDto
            .orElseThrow(() -> new EntityNotFoundException("User not found with ID: " + userId)); // Handle empty optional
}
```